internal/search-history.go

```diff
@@ -0,0 +1,300 @@
+// history. Anything else means the history is ready to go.
+func loadSearchHistory() []string {
+	history, err := loadMoorSearchHistory()
+	if err != nil {
+		log.Infof("Could not load moor search history: %v", err)
+		// IO Error, give up
+		return nil
+	}
+	if history != nil {
+		log.Infof("Loaded %d search history entries from ~/%s", len(history), moorSearchHistoryFileName)
+		return history
+	}
+
+	history, err = loadLessSearchHistory()
+	if err != nil {
+		log.Infof("Could not import less search history: %v", err)
+		return nil
+	}
+	if history == nil {
+		// Neither moor nor less history found, so we start a new history file
+		// from scratch
+		return []string{}
+	}
+
+	log.Infof("Imported %d search history entries from less", len(history))
+	return history
+}
+
+// Try loading search history from ~/.moor_search_history.
+// Returns (nil, nil) if the file doesn't exist. Returns history slice or error.
+func loadMoorSearchHistory() ([]string, error) {
+	lines := []string{}
+	err := iterateFileByLines(moorSearchHistoryFileName, func(line string) {
+		if len(line) > 640 {
+			// Line too long, 640 chars should be enough for anyone
+			return
+		}
+
+		lines = append(lines, line)
+		if len(lines) > maxSearchHistoryEntries {
+			// Throw away the first (oldest) history line, we don't want more
+			// than this
+			lines = lines[1:]
+		}
+	})
+	if errors.Is(err, os.ErrNotExist) {
+		// No history file found, not a problem but no history either, return
+		// nil rather than empty slice
+		return nil, nil
+	}
+
+	if err != nil {
+		return nil, err
+	}
+	return cleanSearchHistory(lines), nil
+}
+
+// Return a new string with any unprintable characters removed
+func withoutUnprintables(s string) string {
+	var builder strings.Builder
+	for _, r := range s {
+		if unicode.IsPrint(r) {
+			builder.WriteRune(r)
+		}
+	}
+	return builder.String()
+}
+
+// File format ref: https://unix.stackexchange.com/a/246641/384864
+func loadLessSearchHistory() ([]string, error) {
+	fileNames := []string{".lesshst", "_lesshst"}
+	for _, fileName := range fileNames {
+		lines := []string{}
+		err := iterateFileByLines(fileName, func(line string) {
+			if !strings.HasPrefix(line, "\"") {
+				// Not a search history line
+				return
+			}
+			if len(line) > 640 {
+				// Line too long, 640 chars should be enough for anyone
+				return
+			}
+
+			lines = append(lines, withoutUnprintables(line[1:])) // Strip leading "
+			if len(lines) > maxSearchHistoryEntries {
+				// Throw away the first (oldest) history line, we don't want more
+				// than this
+				lines = lines[1:]
+			}
+		})
+
+		if errors.Is(err, os.ErrNotExist) {
+			// No such file, try next
+			continue
+		}
+
+		if err != nil {
+			return nil, err
+		}
+
+		return cleanSearchHistory(lines), nil
+	}
+
+	// No history files found, not a problem but no history either, return
+	return nil, nil
+}
+
+func iterateFileByLines(filename string, processLine func(string)) error {
```

Comment:
Could it make sense to re-use `less`'s `LESSHISTFILE` env var if defined and align with its functionality?
```
       LESSHISTFILE
              Name of the history file used to remember search commands and shell commands between invocations  of
              less.   If  set  to  "-"  or  "/dev/null",  a  history  file  is not used.  The default is "$XDG_DA‐
              TA_HOME/lesshst" or "$HOME/.lesshst" on Unix systems, "$HOME/_lesshst" on DOS and  Windows  systems,
              or "$HOME/lesshst.ini" or "$INIT/lesshst.ini" on OS/2 systems.
```

For `XDG_DATA_HOME`, I guess, the https://github.com/adrg/xdg#xdg-base-directory can be referred to (and even used as module) 🤔